src/communicator.jl

```diff
@@ -1,72 +1,143 @@
+    ndev    = length(deviceids)
     comms   = Vector{ncclComm_t}(undef, ndev)
-    devlist = Cint.([i-1 for i in 1:ndev])
-    ncclCommInitAll(comms, ndev, devlist)
-    cs = Communicator.(comms)
-    finalizer(cs) do xs
-        for x in xs
-            ncclCommDestroy(x.handle)
-        end
+    ncclCommInitAll(comms, ndev, deviceids)
+    return map(comms) do ch
+        c = Communicator(ch)
+        finalizer(destroy, c)
     end
-    return cs
 end
+function Communicators(deviceids::AbstractVector{<:Integer})
+    Communicators(Cint[d for d in deviceids])
+end
+function Communicators(devices)
+    Communicators(Cint[deviceid(d) for d in devices])
+end
+
+"""
+    CUDA.deviceid(comm::Communicator)
 
-function device(comm::Communicator)
+The device identifier of the communicator
+
+# External Links
+- [`ncclCommCuDevice`](https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/api/comms.html#ncclcommcudevice)
+"""
+function deviceid(comm::Communicator)
     dev_ref = Ref{Cint}(C_NULL)
-    ncclCommCuDevice(comm.handle, dev_ref)
-    return dev_ref[]
+    ncclCommCuDevice(comm, dev_ref)
+    return Int(dev_ref[])
 end
 
-function Base.size(comm::Communicator)
+
+"""
+    NCCL.size(comm::Communicator) :: Int
+
+The number of communicators in the clique.
+
+# External links
+- [`ncclCommCount`](https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/api/comms.html#ncclcommcount)
+"""
+function size(comm::Communicator)
     size_ref = Ref{Cint}(C_NULL)
-    ncclCommCount(comm.handle, size_ref)
-    return size_ref[]
+    ncclCommCount(comm, size_ref)
+    return Int(size_ref[])
 end
 
+"""
+    NCCL.rank(comm::Communicator) :: Int
+
+The 0-based index of the communicator in the clique.
+
+# External links
+- [`ncclCommUserRank`](https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/api/comms.html#ncclcommuserrank)
+"""
 function rank(comm::Communicator)
     rank_ref = Ref{Cint}(C_NULL)
-    ncclCommUserRank(comm.handle, rank_ref)
-    return rank_ref[]
+    ncclCommUserRank(comm, rank_ref)
+    return Int(rank_ref[])
+end
+
+"""
+    NCCL.abort(comm::Communicator)
+
+Frees resources that are allocated to `comm`. Will abort any
+uncompleted operations before destroying the communicator.
+
+# External links
+- [`ncclCommAbort`](https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/api/comms.html#ncclcommabort)
+"""
+function abort(comm::Communicator)
+    ncclCommAbort(comm)
+    return
+end
+
+
+"""
+    NCCL.default_device_stream(devid::Integer)
+    NCCL.default_device_stream(comm::Communicator)
+
+Get the default stream for device `devid`, or the device corresponding to
+communicator `comm`.
+"""
+function default_device_stream(devid::Integer)
+    state = CUDA.task_local_state!()
+    devidx = devid+1
+    @inbounds if state.streams[devidx] === nothing
+        state.streams[devidx] = CUDA.device!(CUDA.create_stream, devid)
+    else
+        state.streams[devidx]::CuStream
+    end
```

Comment:
This depends on the CUDA internals, so might be something worth adding there?

Reply:
Doesn't seem worth an additional API:

```julia

julia> get_stream(i) = device!(i) do
              stream()
       end

julia> device!(0)
CuDevice(0): NVIDIA H100 PCIe

julia> @benchmark device!(1)
BenchmarkTools.Trial: 10000 samples with 991 evaluations.
 Range (min … max):  41.685 ns … 60.323 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     42.775 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   42.723 ns ±  0.463 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

       ▁▂      ▁     ▅▅▁    █▂                                ▁
  ▅▅▆▇█████▆▆▅██▃▁▄▄▅███▇▆▁▁██▇▄▁▁▁▃▃▁▁▁▁▆▁▁▁▁▁▇▅▁▁▁▁▁▁▁▁▁▁▁▅ █
  41.7 ns      Histogram: log(frequency) by time      44.1 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.
```

Reply:
I changed to CuDevice while querying the stream as shown above; feel free to push additional changes.